pkg/encrypt/server-side.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage
+ * Copyright 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package encrypt
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"errors"
+	"net/http"
+
+	"golang.org/x/crypto/argon2"
+)
+
+const (
+	// sse is the AWS SSE header used for SSE-S3 and SSE-KMS.
+	sse = "X-Amz-Server-Side-Encryption"
+
+	// sseCustomerAlgorithm is the AWS SSE-C algorithm HTTP header key.
+	sseCustomerAlgorithm = sse + "-Customer-Algorithm"
+	// sseCustomerKey is the AWS SSE-C encryption key HTTP header key.
+	sseCustomerKey = sse + "-Customer-Key"
+	// sseCustomerKeyMD5 is the AWS SSE-C encryption key MD5 HTTP header key.
+	sseCustomerKeyMD5 = sse + "-Customer-Key-MD5"
+
+	// sseCopyCustomerAlgorithm is the AWS SSE-C algorithm HTTP header key for CopyObject API.
+	sseCopyCustomerAlgorithm = "X-Amz-Copy-Source-Server-Side-Encryption-Customer-Algorithm"
+	// sseCopyCustomerKey is the AWS SSE-C encryption key HTTP header key for CopyObject API.
+	sseCopyCustomerKey = "X-Amz-Copy-Source-Server-Side-Encryption-Customer-Key"
+	// sseCopyCustomerKeyMD5 is the AWS SSE-C encryption key MD5 HTTP header key for CopyObject API.
+	sseCopyCustomerKeyMD5 = "X-Amz-Copy-Source-Server-Side-Encryption-Customer-Key-MD5"
+)
+
+// PBKDF creates a SSE-C key from the provided password and salt.
+// PBKDF is a password-based key derivation function
+// which can be used to derive a high-entropy cryptographic
+// key from a low-entropy password and a salt.
+type PBKDF func(password, salt []byte) ServerSide
```

Comment:
What about renaming from PBKDF to KeySalter ? I guess it would be easier for users to understand

Reply:
PBKDF is more technically correct here @vadmeste - KeySalter is a very unintuitive name. For Crypto, we can be a bit more expressive.

Reply:
Hm - that is kind of difficult because for example I would not know what a "KeySalter" does and it does not tell me that I deals with passwords two. I see that `PBKDF` is kind of cryptic to an average user - but `PasswordBasedKeyDerivationFunction` is also a WTF and very annoying to write in code.
However the `PBKDF` type is only for advanced user how want to customize password -> key derivation. For example use the `PBKDF2` or `Scrypt`. I guess does users may know what it means...

Reply:
It's only there to provide an upgrade mechanism in the future...

Reply:
Crypto I think we should assume that users know, don't use crypto otherwise you might shoot yourself.

Reply:
The comment helps me understand the acronym PBKDF.